tests/ci/worker/init_runner.sh

```diff
@@ -125,6 +180,20 @@ while true; do
         sleep 15
     else
         echo "Runner is working with pid $runner_pid, nothing to do"
-        sleep 10
+        # The runner does not provide a way to determine, if it runs the job,
+        # neither the way to determine if it just litens. But there should be a
+        # process for Runner.Worker. So if the runner just hangs around for long,
+        # we check if it's fine to let it go
+        if ! pgrep Runner.Worker > /dev/null; then
+            RUNNER_AGE=$(( $(date +%s) - $(stat -c +%Y /proc/"$runner_pid" 2>/dev/null || date +%s) ))
+            echo "The runner is launched $RUNNER_AGE seconds ago and still doesn't have launched Runner.Worker"
+            if (( 60 < RUNNER_AGE )); then
+                echo "Check if the instance should tear down"
+                check-terminating-metadata
```

Comment:
@Felixoid If you have running runner, that has not received the job, I got better results with:
1. Make sure you gracefully stop the runner process
2. Unregister the runner with `./config.sh remove --token ...`

GitHub does not remove your ephemeral runner when you stop the process and it did not process the job.

Not sure about AWS, but make sure the termination just don't kill the runner connection as then GitHub has troubles with determining if it's running, sometimes scheduling job on such runner. Somewhere it's described that GitHub checks for availability every 60 seconds I believe.

Reply:
Thanks, it's useful.

We don't need to unregister the runner, it's done in another [place](https://github.com/ClickHouse/ClickHouse/blob/58da815704c3214e8538eb220103e3be65c736de/tests/ci/ci_runners_metrics_lambda/app.py#L33).

The point regarding stopping the runner seems partially fair. I'll take a look at it.

But it still does not explain the situation, when the GH assigns a job to an unregistered stopped runner.

Reply:
> But it still does not explain the situation, when the GH assigns a job to an unregistered stopped runner.

I did encounter this recently, when I was investigating something, but I could not tell if the runner was unregistered. I cannot reproduce it again. I only know that it was disconnected from the network. Previously it always pointed to the failed graceful shutdown of a runner that did not process a job or was killed right after GitHub would schedule the job on such runner.

It's tricky to determine if a runner got the job scheduled by GitHub or if it's save to shutdown. It would be also possible to run a cron action, to cleanup runners so that runner processes a job and terminates itself afterwards, instead of terminating it from outside. With the tricks I have shared with you I don't encounter such race conditions with couple K daily jobs.

What could also help is to use unique names per ephemeral runner, simple index at the end would do. Then you would be able to determine if GitHub is indeed scheduling the jobs on unregistered stopped runners.

---

> We don't need to unregister the runner, it's done in another [place](https://github.com/ClickHouse/ClickHouse/blob/58da815704c3214e8538eb220103e3be65c736de/tests/ci/ci_runners_metrics_lambda/app.py#L33).

I used runner removal as a fallback to the failed graceful shutdown (there are still requests that may fail, resulting in fake online runner), that's why I do it right away after stopping the runner process. From the code you sent it feels like it's an async process in your case.

Reply:
> It's tricky to determine if a runner got the job scheduled by GitHub or if it's save to shutdown

This is a terrible architecture issue, TBH. For like, GitHub actions look like a pet project to me, that surprisingly became a product, without considering edge cases for intensive usage. Like, I check for `Runner.Worker`, it's omitted, I kill it, but then I have

```
The runner is launched 60 seconds ago and still doesn't have launched Runner.Worker
Got runner pid 2076809
Runner is working with pid 2076809, checking the metadata in background
The runner is launched 65 seconds ago and still doesn't have launched Runner.Worker
Check if the instance should tear down
The instance in ASG status Terminating:Wait
/home/ubuntu/actions-runner/run-helper.sh: line 36: 2076809 Killed                  "$DIR"/bin/Runner.Listener run $*
Exiting with unknown error code: 137
Exiting runner...
# Runner removal
Failed: Removing runner from the server
Runner "i-06ef5a2b94fc199f8" is still running a job"
The runner is marked as "Terminated" by the autoscaling group, we are terminating
```

It's nonsense!

> What could also help is to use unique names per ephemeral runner, simple index at the end would do. Then you would be able to determine if GitHub is indeed scheduling the jobs on unregistered stopped runners.

It has a drawback since we compare the instance ID to the runner name. On the other hand, it actually looks promising. Unless GitHub would even ignore the deleted runner, lol.

> From the code you sent it feels like it's an async process in your case.

True. The instant removal makes sense, although it's broken for `kill -9`. But it seems fair to me.